Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp

```diff
@@ -107,6 +132,43 @@ void ImageBufferSkiaAcceleratedBackend::putPixelBuffer(const PixelBuffer& pixelB
     m_surface->writePixels(pixmap, 0, 0);
 }
 
+#if USE(NICOSIA)
+RefPtr<GraphicsLayerContentsDisplayDelegate> ImageBufferSkiaAcceleratedBackend::layerContentsDisplayDelegate() const
+{
+    return m_layerContentsDisplayDelegate;
+}
+
+void ImageBufferSkiaAcceleratedBackend::swapBuffersIfNeeded()
+{
+    if (!PlatformDisplay::sharedDisplayForCompositing().skiaGLContext()->makeContextCurrent())
+        return;
+
+    RELEASE_ASSERT(m_contentLayer);
+
+    auto* grContext = PlatformDisplay::sharedDisplayForCompositing().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+    grContext->flushAndSubmit(m_surface.get(), GLFence::isSupported() ? GrSyncCpu::kNo : GrSyncCpu::kYes);
+
+    auto texture = SkSurfaces::GetBackendTexture(m_surface.get(), SkSurface::BackendHandleAccess::kFlushRead);
+    ASSERT(texture.isValid());
+    GrGLTextureInfo textureInfo;
+    bool retrievedTextureInfo = GrBackendTextures::GetGLTextureInfo(texture, &textureInfo);
+    ASSERT_UNUSED(retrievedTextureInfo, retrievedTextureInfo);
+
+    auto info = m_surface->imageInfo();
+    IntSize textureSize(info.width(), info.height());
+    auto copiedTexture = BitmapTexture::create(textureSize, BitmapTexture::Flags::SupportsAlpha);
```

Comment:
While this works, creating a texture for each frame is not efficient. We should have 2 BitmapTextures precreated that we're swapping as write and read buffers. During each call, we update the contents of the write one with the contents of the canvas, swap the textures so the write one becomes the read one, and push the read one to the proxy. But this can be done later in a different commit.